src/components/Todos/Todos.tsx

```diff
@@ -0,0 +1,156 @@
+import { useEffect, useRef, useState } from 'react';
+import { useTodoContext } from '../context/TodoContext';
+
+export const Todos = () => {
+  const [editingTodoId, setEditingTodoId] = useState<number | null>(null);
+  const [editingTitle, setEditingTitle] = useState('');
+  const renameInputRef = useRef<HTMLInputElement | null>(null);
+  const {
+    toggleTodoStatus,
+    updateTodoTitle,
+    deleteTodo,
+    headerInputRef,
+    filteredTodos,
+  } = useTodoContext();
+
+  const handleSubmit = (event: React.FormEvent) => {
+    event.preventDefault();
+
+    if (editingTodoId !== null) {
+      if (editingTitle.trim() === '') {
+        deleteTodo(editingTodoId);
+      } else {
+        updateTodoTitle(editingTodoId, editingTitle);
+      }
+    }
+
+    setEditingTodoId(null);
+
+    if (headerInputRef.current) {
+      headerInputRef.current.focus();
+    }
+  };
+
+  const handleDoubleClick = (id: number, currentTitle: string) => {
+    setEditingTodoId(id);
+    setEditingTitle(currentTitle);
+  };
+
+  useEffect(() => {
+    if (editingTodoId !== null && renameInputRef.current) {
+      renameInputRef.current.focus();
+    }
+
+    const handleClickOutsideForm = (event: MouseEvent) => {
+      if (
+        editingTodoId !== null &&
+        renameInputRef.current &&
+        !renameInputRef.current.contains(event.target as Node)
+      ) {
+        if (editingTitle.trim() === '') {
+          deleteTodo(editingTodoId);
+        } else {
+          updateTodoTitle(editingTodoId, editingTitle);
+        }
+
+        setEditingTodoId(null);
+
+        setTimeout(() => {
+          if (headerInputRef.current) {
+            headerInputRef.current.focus();
+          }
+        }, 0);
+      }
+    };
```

Comment:
You don't need this logic. You can just simply listen for onBlur event on the input